File: CrackTheCodeInterview/src/practiceQuestions/GenerateParentheses.java
package practiceQuestions;

import java.util.ArrayList;
import java.util.List;

/*
 * question: https://leetcode.com/problems/generate-parentheses/
 * 
 * explanation:
 * https://algorithms.tutorialhorizon.com/generate-all-valid-parenthesis-strings-of-length-2n-of-given-n/
 * https://www.youtube.com/watch?v=sz1qaKt0KGQ
 */
public class GenerateParentheses {

	List<String> res;

	public List<String> generateParenthesis(int n) {
		res = new ArrayList<>();
		addParenthesis(n, n, "");

		return res;
	}

	void addParenthesis(int open, int close, String string) {
		if (open == 0 && close == 0) {
			res.add(string);
		}

		if (close < open) {
			return;
		}

		if (open > 0) {
			addParenthesis(open - 1, close, string + "(");
		}

		if (close > 0) {
			addParenthesis(open, close - 1, string + ")");
		}
	}
}

File: go/generate_parentheses.go
package main

// question: https://leetcode.com/problems/generate-parentheses/
func generateParenthesis(n int) []string {
	res := make([]string, 0)
	buildParenthesis("", n, n, &res)
	return res
}

func buildParenthesis(subString string, left int, right int, res *[]string) {
	if left > right {
		return
	}

	if left > 0 {
		buildParenthesis(subString+"(", left-1, right, res)
	}

	if right > 0 {
		buildParenthesis(subString+")", left, right-1, res)
	}

	if left == 0 && right == 0 {
		*res = append(*res, subString)
		return
	}
}
